src/development/cplusplus-codeguide.rst

```diff
@@ -0,0 +1,296 @@
+
+In general however we recommend Modern CMake (minimum version of 3.0). A lot of legacy
+projects will have used 2.8 - but we cannot recommend that new projects use
+this. It is no longer maintained and does not contain many features of modern
+(v3+ CMake). As far as we are aware CMake v3.0 is backward compatible
+with 2.8, so if you are onboarding code that uses 2.8 we strongly suggest you upgrade. 
+
+.. todo:: Maybe add some more advice? use targets not global settings etc ...
+
+Including the version number
+""""""""""""""""""""""""""""
+There is a template configuration file that demonstrates how to pull in the
+version number into the code base. We do this in a few  steps. First the Semantic
+Version number is held in plain text in version.txt. This is done so that it is
+easy to "bump" the version number on code changes. The second stage is the the
+CMakeLists.txt file reads this file and creates an internal version number
+using its contents. Finally the configuration file is used to generate a header
+file containing the version number that can be included by the code.
+
+Including External Projects
+"""""""""""""""""""""""""""
+We have including the GoogleTest framework as an external project, instead of a
+dependency to demonstrate one way to use external projects. 
+
+.. warning:: We have used a git submodule to do this. There may be a formal recomendation that git subtrees be used instead.
+
+Runtime Dependencies
+""""""""""""""""""""
+
+.. todo:: Add a simple description of dependency and version listing
+
+We consider that there are two principles of dependency management we would dictate:
+
+1) Dependency installation should be unambigigous and repeatable. This requires dependencies to be listed, together with a minimum, or required version number. 
+
+2) Dependencies should be project specific, there should be no reason for developers to pollute their environment just to build your package.
+
+We do not advocate a solution here - there is not a dominant package management system for C++. Indeed you could just pull and build all your deopendencies via external projects as the our GoogleTest example. Another option would simply be to list the dependencies in a text file *dependencies.txt* and allow the developer to use their preferred pacakge management system to install them.
+
+Coding Style & Conventions
+--------------------------
+
+We are not advocating that software be restructured and rewritten before
+on-boarding - However we recommend that new software follow `The cplusplus Core
+Guidelines <http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines>`_.
+
+The clang/llvm compiler tools have an extension which can provide some direct
+criticism of your code for stylistic errors (and even automatically fix them). For example in our lint step we
+suggest you run:: 
+
+    run-clang-tidy -checks='cppcoreguidelines-*,performance-*,readibility-*,modernize-*,misc-*,clang-analyzer-*,google-*'
+
```

Comment:
can we add a ClangTidy.cmake module that provides a "make lint" target that does this automatically?

Reply:
I've added a target. It does depend upon the compilation database being in existence tho otherwise it will fall over.